doc/developer/design/20231102_read_holds.md

```diff
@@ -0,0 +1,154 @@
+It is an error if that time is less than any object's logical read frontier.
+If `TO <time>` is not specified, the new time is the least common physical read frontier of all objects.
+The `HOLD` releases its current read hold and aquires a new read hold at the new time.
+
+The `ERROR_IF_OLDER_THAN` option (required, but defaults to `'24h'`) is a duration.
+If difference between the least common physical read frontier of the objects and the `HOLD` time is greater than `ERROR_IF_OLDER_THAN`, the `HOLD`'s read hold is released and the `HOLD` enters an error state.
+(Transactions and `SUBSCRIBE` are not affected, as in the `ALTER HOLD ADVANCE` case.)
+In the error state, `ALTER HOLD ADVANCE` will return an error.
+`SELECT` and `SUBSCRIBE` `AS OF` a time that is between the erroring `HOLD`'s time and the least common physical read frontier will return a helpful error message that the `HOLD` errored because it was not advanced.
+
+It is an error to `DROP` an object that is part of a `HOLD` unless the `DROP` is `CASCADE`.
+If the `DROP` is `CASCADE`, any `HOLD`s on the dropped object are also dropped (even if the `HOLD` is also over other objects).
+
+### Monitoring
+
+To monitor `HOLD`s, a new `mz_catalog.mz_holds` table will be maintained with structure:
+
+- `id text`: the `HOLD`'s id
+- `database_id text`
+- `schema_id text`
+- `name text`
+- `on text[]`: an array of the object ids in the `HOLD`
+- `at uint8`: the `HOLD`'s time
+
+### Implementation of time tracking
+
+The `HOLD` object has an associated time that must be written down somewhere on creation then updated on advancement.
+This will use a persist shard to back it, and is thus similar to a 1-row table.
+As a result of not storing it in the catalog (see Alternatives for the motivation), the `AT <time>` syntax must not be part of the `create_sql` of the `HOLD` object and must be purified away.
+On coordinator bootstrap, a `HOLD` will do a 1-time fetch of its state from the shard to resume its operation.
+
+## Minimal Viable Prototype
+
+An MVP exists at https://github.com/MaterializeInc/materialize/pull/22976.
+
+## Alternatives
+
+- Instead of supporting multiple objects, we could support a single object and enforce the `HOLD` is created in its schema.
+Or, we could support multiple objects, but they must all be in the same schema and the `HOLD` must be in that schema as well.
+
+- These could be called `READ HOLD`s instead of `HOLD`s.
+
+- `SUBSCRIBES` themselves could be made durable, or we introduce `CREATE SUBSCRIPTION` that would similarly achieve this goal ((proposal)[https://github.com/MaterializeInc/materialize/issues/17541#issuecomment-1569340739]).
+
+- Allow users to configure the compaction window of objects.
+This has problems: the window could be too short for recovery or too long and overuse resources, OOMing a cluster.
+Having users routinely advance a `HOLD` should scale the extra resource use to exactly how far behind a user is.
```

Comment:
These problems seem like the exist with read holds as well. A compaction window seems 100x less likely to result in OOMing a cluster than a hold that someone needs to remember to refresh, or when an automated refreshing task experiences an ongoing fault and can't be run. An ongoing automated refreshing task sounds very much like a compaction window we ask someone else to run for us. It seems like the positive, that the amount of history adapts to the use, is specifically tailored to subscribes, but I would want to make sure our compaction controls are general enough for multiple use cases.

Reply:
> Thoughts about also removing manual compaction completely? Does it add too much complication for users, or are its memory benefits worth it?

I think the memory benefits are worth it! But, agreed that not every user will want the complexity of holds. That's why I've been pushing for re-enabling user-configurable compaction windows as part of this bundle of work: https://github.com/MaterializeInc/materialize/issues/16701. Is that still the plan? 

IMO, the reason to _do_ holds is so that you can drive advancement faster in the happy case. For users who just want a grace period where they can reconnect and resume where they left off, I think just setting a 5m (15m? 60m?) logical compaction window is a good way to go.

Reply:
Ah I must've missed/dropped that we wanted to include re-enabling user-configurable compaction windows more generally in this first phase of work. I had originally imagined us releasing read holds first to unblock durable subscribes for the reverse ETL use case, then moving on to general purpose user controlled compaction windows.
Do we know what the product use cases are specifically that we're supporting with them?

But if it's more efficient to design & build both together then I'm supportive.

I'm also fine with supporting _only_ automatically advancing read holds (which is basically the same thing as custom compaction afaiu) as Matt suggested, if that narrows scope. The user will already need to provision their cluster for the worst-case scenario, so perhaps the manual advancement is not that useful?

Reply:
> Do we know what the product use cases are specifically that we're supporting with them?

Potentially similar product use cases to read holds! It may be that rETL tools would prefer to simply require a custom compaction window of, say, 1hr on all the models they read from, and on their side they will try very hard to recover from any failures in 1hr, accepting responsibility for processing a new snapshot.

I would think of this all as discovery work. We don’t know exactly how users will use these features ahead of time, and they’re in the weeds enough that we can’t meaningfully engage with users until we have working versions. It’s been realllly hard to get anyone to engage with read holds in the abstract. But I’m confident enough that both of these features are sensible and valuable that I’d like to advocate for introducing them both in private preview, and then we can iterate on them behind private preview as users show up with real use cases for each.

FWIW, if we have to choose on ordering, I’d rather re-add support for user configurable compaction windows _first_. (I was hoping we could do both concurrently, as read holds have some thorny design problems that configurable compaction windows do not have.) Configurable compaction windows are substantially simpler than read holds, and could potentially unblock rETL tools faster than building full support for read holds. The rETL tools could get started building resumable SUBSCRIBEs based on relations with, say, 1hr compaction windows, and then later, once we’ve shipped read holds, could learn to use read holds to substantially reduce resource requirements when things are keeping up. 

> I'm also fine with supporting only automatically advancing read holds (which is basically the same thing as custom compaction afaiu) as Matt suggested, if that narrows scope. The user will already need to provision their cluster for the worst-case scenario, so perhaps the manual advancement is not that useful?

The raison d’etre of read holds is that they support manual advancement! If we don’t support manual advancement of read holds, IMO we should not build read holds, and just do user-configurable compaction windows. At least, I can’t see what read holds that only support manual advancement offer over user-configurable compaction policies.

That said: I think read holds are an incredibly valuable tool, and one I’m very bullish on building. I’d like to ultimately support them with ~no restriction on the maximum value for `MAX LAG`. You wouldn’t actually need to provision for the worst case scenario, I don’t think. You’d provision for the steady state, and then if your app had an issue that meant it failed to keep up with the upstream system and a bunch of historical revisions piled up, after bringing your app back online, you could temporarily size up your cluster to get through the backlog, then size it back down.